### PR TITLE
CLOUDP-80265: Implement Cluster update logic

### PR DESCRIPTION
**Pull request by vasilevp**

### CODE REVIEW

File: pkg/controller/atlascluster/cluster.go [hunk 1]
```diff
@@ -38,17 +44,66 @@ func ensureClusterState(wctx *workflow.Context, connection atlas.Connection, pro
...
+		c, _, err = client.Clusters.Update(ctx, project.Status.ID, cluster.Spec.Name, spec)
+		if err != nil {
+			return c, workflow.Terminate(workflow.ClusterNotCreatedInAtlas, err.Error())
+		}
+
+		return c, workflow.InProgress(workflow.ClusterNotUpToDate, "cluster update started")
 
 	case "CREATING":
-		return c, workflow.InProgress(workflow.ClusterCreating, "Cluster is provisioning")
+		return c, workflow.InProgress(workflow.ClusterCreating, "cluster is provisioning")
```
Comment (antonlisovenko) [inline pkg/controller/atlascluster/cluster.go:66]: [nit] why lowercase? I believe all status fields are in common start from Uppercase

File: pkg/controller/atlascluster/cluster.go [hunk 2]
```diff
@@ -38,17 +44,66 @@ func ensureClusterState(wctx *workflow.Context, connection atlas.Connection, pro
...
+		return c, workflow.InProgress(workflow.ClusterUpdating, "cluster is updating")
 
 	// TODO: add "DELETING", "DELETED", handle 404 on delete
 
 	default:
 		return c, workflow.Terminate(workflow.Internal, fmt.Sprintf("unknown cluster state %q", c.StateName))
 	}
 }
+
+// clusterMatchesSpec will merge everything from the Spec into existing Cluster and use that to detect change
```
Comment (antonlisovenko) [inline pkg/controller/atlascluster/cluster.go:78]: [nit] a bit more information is welcome (e.g. why we are doing this and cannot compare objects directly - as the cluster from Atlas may have default/deprecated fields)

**Review by antonlisovenko — APPROVED** [review]
Looks great, as discussed the edge cases ("paused" etc) can be addressed later (please create a ticket!)

Also please update the description with the examples of the update cluster command

🥇

File: pkg/controller/atlascluster/cluster.go (hunk 1 above)
Comment (vasilevp) [reply pkg/controller/atlascluster/cluster.go:66]: most of our status fields are `err.Error()`, which, by convention, is lowercase

Comment (antonlisovenko) [reply pkg/controller/atlascluster/cluster.go:66]: ah I see. Then it makes sense to uppercase them before writing to the status.
Ok then 👍

Comment (vasilevp) [reply pkg/controller/atlascluster/cluster.go:66]: we could also prepend something like `"Error: " + err.Error()` to simplify things

**Comment by vasilevp** [comment]
POC update sequence:
```sh
$ kubectl apply -f bin/cluster.yaml -f bin/project.yaml # enabled cloud backups in cluster.yaml
```
```yaml
atlascluster.atlas.mongodb.com/crdtest2 configured
atlasproject.atlas.mongodb.com/crdtest2 unchanged
```
```sh
$ kubectl get atlasclusters.atlas.mongodb.com crdtest2 -o yaml | yq r - status # after 10s
```
```yaml
conditions:
- lastTransitionTime: "2021-01-19T20:17:28Z"
  message: cluster is updating
  reason: ClusterUpdating
  status: "False"
  type: ClusterReady
connectionStrings:
  standard: mongodb://crdtest2-shard-00-00.0ey7x.mongodb.net:27017,crdtest2-shard-00-01.0ey7x.mongodb.net:27017,crdtest2-shard-00-02.0ey7x.mongodb.net:27017/?ssl=true&authSource=admin&replicaSet=atlas-7sih41-shard-0
  standardSrv: mongodb+srv://crdtest2.0ey7x.mongodb.net
mongoDBVersion: 4.2.11
mongoURIUpdated: "2021-01-15T10:51:45Z"
stateName: UPDATING
```
```sh
$ kubectl get atlasclusters.atlas.mongodb.com crdtest2 -o yaml | yq r - status # after 30s
```
```yaml
conditions:
- lastTransitionTime: "2021-01-19T20:17:38Z"
  status: "True"
  type: ClusterReady
connectionStrings:
  standard: mongodb://crdtest2-shard-00-00.0ey7x.mongodb.net:27017,crdtest2-shard-00-01.0ey7x.mongodb.net:27017,crdtest2-shard-00-02.0ey7x.mongodb.net:27017/?ssl=true&authSource=admin&replicaSet=atlas-7sih41-shard-0
  standardSrv: mongodb+srv://crdtest2.0ey7x.mongodb.net
mongoDBVersion: 4.2.11
mongoURIUpdated: "2021-01-15T10:51:45Z"
stateName: IDLE
```

**Comment by vasilevp** [comment]
@antonlisovenko FYI ^

**Pull request merged**